pymc3/distributions/multivariate.py

```diff
@@ -1790,119 +1805,83 @@ class KroneckerNormal(Continuous):
-        else:
-            self._cov_type = "evd"
-            self._setup_evd(evds)
-
-    def _setup_evd(self, eigh_iterable):
-        self._isEVD = True
-        eigs_sep, Qs = zip(*eigh_iterable)  # Unzip
-        self.Qs = list(map(at.as_tensor_variable, Qs))
-        self.QTs = list(map(at.transpose, self.Qs))
-
-        self.eigs_sep = list(map(at.as_tensor_variable, eigs_sep))
-        self.eigs = kron_diag(*self.eigs_sep)  # Combine separate eigs
-        if self.is_noisy:
-            self.eigs += self.sigma ** 2
-        self.N = self.eigs.shape[0]
-
-    def _setup_random(self):
-        if not hasattr(self, "mv_params"):
-            self.mv_params = {"mu": self.mu}
-            if self._cov_type == "cov":
-                cov = kronecker(*self.covs)
-                if self.is_noisy:
-                    cov = cov + self.sigma ** 2 * at.identity_like(cov)
-                self.mv_params["cov"] = cov
-            elif self._cov_type == "chol":
-                if self.is_noisy:
-                    covs = []
-                    for eig, Q in zip(self.eigs_sep, self.Qs):
-                        cov_i = at.dot(Q, at.dot(at.diag(eig), Q.T))
-                        covs.append(cov_i)
-                    cov = kronecker(*covs)
-                    if self.is_noisy:
-                        cov = cov + self.sigma ** 2 * at.identity_like(cov)
-                    self.mv_params["chol"] = self.cholesky(cov)
-                else:
-                    self.mv_params["chol"] = kronecker(*self.chols)
-            elif self._cov_type == "evd":
+                eigh_iterable = eigh_map
                 covs = []
-                for eig, Q in zip(self.eigs_sep, self.Qs):
+                for eig, Q in zip(eigs_sep, Qs):
                     cov_i = at.dot(Q, at.dot(at.diag(eig), Q.T))
                     covs.append(cov_i)
                 cov = kronecker(*covs)
-                if self.is_noisy:
-                    cov = cov + self.sigma ** 2 * at.identity_like(cov)
-                self.mv_params["cov"] = cov
+                if is_noisy:
+                    cov = cov + sigma ** 2 * at.identity_like(cov)
+                chol = cholesky(cov)
+            else:
+                chol_diags = list(map(at.diag, chols))
+                cov = kronecker(*chols)
+            # Check if this is the correct operation
+            cov = quaddist_matrix(cov, chol)
+        else:
+            eigh_iterable = evds
+            covs = []
+            for eig, Q in zip(eigs_sep, Qs):
+                cov_i = at.dot(Q, at.dot(at.diag(eig), Q.T))
+                covs.append(cov_i)
+            cov = kronecker(*covs)
+            if is_noisy:
+                cov = cov + sigma ** 2 * at.identity_like(cov)
 
-    def random(self, point=None, size=None):
+        mu = at.as_tensor_variable(mu)
+
+        chols = at.as_tensor_variable(chols)
+
+        # mean = median = mode = mu
+        return super().dist([mu, cov, chols, eigh_iterable], **kwargs)
```

Comment:
@brandonwillard we discussed the possibility of using a parametrization of just `super().dist([mu, *cov], **kwargs)`. That will be fine for `rng_fn` but the extra args required for `logp` must be passed if we are to support `evd` there.

Reply:
Another, not very pretty option might be to use a specific `_logp` dispatcher, which extracts the extra non-numeric variables from the `tag` and passes them on to the `logp` function